Pass subprocess return status to W&B using subprocess.Popen

jondoglover opened the pull request:
- W&B should now almost always correctly display whether a run failed or was successful and also display relevant error information in the W&B log. 
- subprocess.Popen and subprocess.run were both considered. Although cleaner and simpler, subprocess.run often fails to capture the stack trace, so subprocess.Popen is a better option.

kylebgorman commented on examples/wandb_sweeps/sweep.py:
```diff
@@ -29,22 +29,48 @@ def train_sweep(
-        subprocess.check_call(argv)
-    except subprocess.CalledProcessError as error:
-        logging.error("Subprocess error: %s", error)
+        process = subprocess.Popen(argv, stderr=subprocess.PIPE)
```
The try/except part is no longer necessary. The outer call is wrapped in a try/except.

DanielYakubov commented on examples/wandb_sweeps/sweep.py:
```diff
@@ -29,22 +29,48 @@ def train_sweep(
+        return process.returncode
+
+    except Exception:
+        error_msg = f"Full traceback: {traceback.format_exc()}"
+        logging.error(error_msg)
```
I think it's simpler to just one line this (it's self documenting that it's an error message)

logging.error(f"Full traceback: {traceback.format_exc()}")

Same comment for 65,66

jondoglover commented:
Thanks for all the suggestions! I’ve implemented most of the changes, though I kept the while loop since subprocess.Popen is non-blocking. Otherwise, I simplified it significantly for better clarity and efficiency. Please let me know if there’s anything else that should be adjusted.

kylebgorman commented on examples/wandb_sweeps/sweep.py:
```diff
@@ -41,10 +42,19 @@ def run_sweep(argv: List[str]) -> None:
-    except subprocess.CalledProcessError as error:
-        logging.error("Subprocess error: %s", error)
+    process = subprocess.Popen(argv, stderr=subprocess.PIPE, text=True)
+    # subprocess.Popen is nonblocking, so the while loop waits for
```
So, creating a `Popen` instance doesn't block, but looping over it in the idiomatic way I recommend does.

I wasn't completely positive I was right about this, so I made a toy example. The subprocess waits 10 seconds then prints a line to stderr. The wrapper calls that subprocess using `Popen`. I think you'll see it works just fine without this additional complexity (see instructions there):

https://gist.github.com/kylebgorman/191c4496f5d0c3b98c025b631b470b2c

Were this not blocking, the loop in the wrapper would have terminated long before the 10 seconds are up. 

Of course it may be the case that there is some reason my toy example is non-representative of what you're trying to do here, and if so let me know. But please take a look...everything else looks fine.

jondoglover replied:
@kylebgorman Thanks for the gist! I've extended slow.py in the gist to return a code, but it seems the return code isn’t being captured. I suspect that this is due to a gap between when the pipe closes and when the process actually ends and returns a code.

With subprocess.Popen, it seems we need either subprocess.poll() or subprocess.communicate() to ensure the subprocess is properly terminated. While subprocess.communicate() feels more idiomatic, my sample runs show that it doesn’t consistently capture the stack trace. If we decide to use subprocess.communicate(), we might consider using subprocess.run() instead for simplicity.

Additionally, I was able to simplify the while loop to make it more concise and idiomatic. Between these approaches, which would you recommend? Thanks again! =)

kylebgorman replied:
That's not a problem either; you just call `process.wait()` at the end of the loop, which returns the exit code. I updated the example to make sure I wasn't crazy. `while True:` and `.readline` are both code smells, so I recommend at least trying to adapt this design.

kylebgorman reviewed (commented):
One last nit. Thanks Jonathan.

kylebgorman commented on examples/wandb_sweeps/sweep.py:
```diff
@@ -41,10 +41,12 @@ def run_sweep(argv: List[str]) -> None:
+    process = subprocess.Popen(argv, stderr=subprocess.PIPE, text=True)
+    for line in process.stderr:
+        logging.info(line.rstrip())
+    return_code = process.wait()
```
My last nit seems to have been swallowed by GitHub so I'm trying again. I see why you were confused!

So right now the `info` log is just the stderr from the subprocess, and that seems fine to me. I think adding the explicit exit code to the log is not necessary, since I confirmed (by looking at the W&B source) that it already logs the exit code (which is why `finish` takes that as an argument) at a higher level. So I'd remove line 48 and combine lines 47 and 49 into one, and that's my last ask. 

In general---and reasonable people may disagree with this---I prefer:

```python
finalize(foo(...))
```

to

```python
bar = foo(...)
finalize(bar)
```

except when:

1. the right-hand side (RHS) expression I'm abbreviating `foo(...)` is unusually long or complex, or 
2. (relatedly) it is hard to reason from context what `foo(...)` is, exactly, and the variable name gives the reader some context, or
3. the RHS will be used again in the same scope so it's sensible to save it somewhere.
4. (relatedly) there is some reason to override the variable.

Here I'd say it may not be immediately obvious to the reader that `wait` returns the exit code (even if that is the documented API for this built-in module), so (2) is possibly applicable, but

```python
wandb.finish(exit_code=process.wait())
```

makes it reasonably obvious what's going on and negates the benefit.

kylebgorman reviewed (approved):
LGTM, thanks!
